src/shared/sleep-config.c

```diff
@@ -141,6 +142,147 @@ int read_battery_capacity_percentage(void) {
+        _cleanup_free_ char *p = NULL, *battery_id = NULL;
+        sd_id128_t machine_id, product_id;
+        int r;
+
+        assert(ret_hash);
+
+        r = get_battery_identifiers(&battery_id);
+        if (r < 0)
+                return log_debug_errno(r, "Failed to get battery identifier: %m");
+
+        r = sd_id128_get_machine(&machine_id);
+        if (r < 0)
+                return log_debug_errno(r, "Failed to get machine ID: %m");
+
+        r = id128_get_product(&product_id);
+        if (r < 0)
+                return log_debug_errno(r, "Failed to get product ID: %m");
+
+        p = strjoina(battery_id, ":", SD_ID128_TO_STRING(machine_id), ":", SD_ID128_TO_STRING(product_id));
+        if (!p)
+                return log_oom_debug();
+        /* add hash function when everything is working */
+        *ret_hash = TAKE_PTR(p);
+
+        return 0;
+}
+
+static int battery_discharge_rate_in_range(int battery_discharge_rate) {
+        if (battery_discharge_rate <= 0 || battery_discharge_rate >= 200)
+               return 0;
+               /* battery discharge rate is not in a valid range */
+        log_debug("Battery discharge rate is in valid range");
+        return 1;
+}
+
+int get_battery_discharge_rate(void) {
+        _cleanup_free_ char *hash_id_discharge_rate = NULL;
+        const char *p = NULL, *filepath = NULL;
+        char *current_hash_id = NULL;
+        char *stored_hash_id = NULL, *stored_discharge_rate = NULL;
+        int discharge_rate, r;
+
+        filepath = "/var/lib/systemd/sleep/battery_discharge_percentage_rate_per_hour";
+
+        r = read_one_line_file(filepath, &hash_id_discharge_rate);
+        log_debug("hash_id_discharge_rate is %s", hash_id_discharge_rate);
+        if (r < 0)
+               return log_debug_errno(r, "Failed to read discharge rate from %s: %m", filepath);
+        p = hash_id_discharge_rate;
+
+        r = get_system_battery_identifier_hash(&current_hash_id);
+        log_debug("system_battery_identifier is %s", current_hash_id);
+        if (r < 0)
+               return log_debug_errno(r, "Failed to generate system battery identifier hash: %m");
+
+        /* extract the string in file, split it, validate the hash and then return discharge rate */
+
+        /* if (!first_word(hash_id_discharge_rate, current_hash_id))
+         *       return log_debug_errno(r, "Current system and battery identifier do not match the stored identifier: %m"); */
+
+        r = extract_many_words(&p, " ", 0, &stored_hash_id, &stored_discharge_rate, NULL);
+        log_debug("stored hash_id is %s and stored discharge_rate is %s", stored_hash_id, stored_discharge_rate);
+        if (r < 0)
+               return log_debug_errno(r, "Failed to parse hash_id and discharge_rate read from %s location: %m", filepath);
+
+        if (!strcmp_ptr(current_hash_id, stored_hash_id))
+               return log_debug_errno(r, "Current system and battery identifier do not match the stored identifier: %m");
```

Comment:
returning `r` here is not the right error code, it should return ENOENT - also it should write/read multiple lines until the matching hash is found (or ENOENT if it cannot be found)

Reply:
Once the single line starts to work, i will add multiple line looping

Reply:
yes that's all good

Reply:
but it's still returning `r` which will not contain any error, it needs to return ENOENT

Reply:
i commented that when inserted first_word. Weird it still shows uncommented here. In code i have pushed, its commented. https://github.com/cerebro1/systemd/blob/week1/src/shared/sleep-config.c#L222
But will be replacing first_word now so will take care of it when adding `streq`